### ISSUE TITLE
Current dll linking has timeouts on Windows

### ISSUE THREAD

**Issue by danielpclark** [danielpclark/rutie#48]
I found a way to get the Windows test suite to run by generating a lib file for the Ruby library's dll but many to most of the tests now time out in windows.

It's important to note that this issue didn't exist when Windows CI testing had been working previously so once the solution is found it should just work on Windows.

Even though `libgmp-10.dll` is in the path the linker is having issues linking with it.

**Comment by danielpclark** [comment]
I'm thinking of giving pkg-cfg a try on Windows if see if it just works with that rather than the custom lib generation I have now.

**Comment by danielpclark** [comment]
Actually I might have the fix...

**Comment by danielpclark** [comment]
Okay.  So using the Visual Studio debugger I'm finding that everything that's failing is failing the same way.  It looks like the `const_get` method in Rutie is trying to access invalid memory only on Windows.  The debugger shows the following message.

```
Unhandled exception at 0x0000000068034E74 (x64-mscvrt-ruby250.dll) in rust_out.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF
```

And it's highlighting

```rust
pub fn const_get(klass: Value, name: &str) -> Value {
    unsafe { class::rb_const_get(klass, symbol::internal_id(name)) }
}
```

as the problem area.

Maybe something between 64bit/32bit isn't quite right on Windows.  ¯\_(ツ)_/¯

**Comment by pravic** [comment]
Hi, just walked by.

By linking you mean something different, a Ruby-specific, right? It was also mentioned [in the README](https://github.com/danielpclark/rutie#error-while-loading-shared-libraries-librubyso-cannot-open-shared-object-file-no-such-file-or-directory).

**Comment by danielpclark** [comment]
Thanks @pravic , but the tests in Rutie are all run from the Rust side and not from Ruby.  What I mean by linking is the Rust compiler **rustc**  has the GMP library linking the dll during compilation and has occasional issues with this library.  So this issue isn't Ruby-specific.

Copying `libgmp-10.dll` to one of `gmp-10.dll` or `gmp.dll` and placing it in the deps dir may solve the linking issue for this dll.  But it doesn't solve the timeouts.

**Comment by pravic** [comment]
It's still not clear.

How to reproduce?  Just a `cargo test`? Sorry for silly questions,  as a Windows citizen I'd can look, but I need some walk through.

**Comment by danielpclark** [comment]
Yes.  Either toolchain should produce the same result.  The TravisCI suite and myself are using the msvc toolchain.  `cargo test` should be all you need to reproduce.

Primarily it will say something like `rust_out.exe` has timed out.  If a copy of the GMP dll hasn't been made as I indicated above you may get a few dialogs popping up about it.  If you have Visual Studio and you're on Windows 10 the time out error will show something like a “Debug” button which is where I gathered the info above from.  You can also look under TravisCI for the full trace output: https://travis-ci.org/danielpclark/rutie/jobs/448965441

**Comment by pravic** [comment]
Well..

First of all, Ruby is compiled by MinGW by default, so it's quite hard to debug. I've recompiled it with MSVC (`win32\configure.bat --target=x64-mswin64 --disable-install-doc --disable-rubygems --without-ext=ripper`).

Then I tried to debug and it has crashed inside `_pioinfo(int fd)` in win32.c, because `__pioinfo` was not initialized, because `set_pioinfo_extra` wasn't called. 

However, after I added `ruby_sysinit` it still crashed in `rb_ec_tag_jump` because of null `ec->tag`. The exception raised after `Class::new("SomeClass", None)` with message "wrong argument type Integer (expected Class)".

**Comment by danielpclark** [comment]
`ec` is the execution context and perhaps one can't exist before `vm::ruby_init()`?  `nil` should be a valid parameter for an alternative execution context, but not null.

Each `VALUE` object in Ruby has an [internal type associated with it](https://github.com/danielpclark/rutie/blob/689264718e942f67f873cd8a78cfe1cc83f6053b/src/rubysys/value.rs#L45-L79).  It sounds like the `ValueType` aka `ruby_value_type` ended up being `0x15` instead of `0x02` which would possibly explain the `Integer` message.  Although I doubt that is useful to know.

My thoughts on how I'd possibly try to sort this out are of two mindsets. Since older versions of this project did work with older versions of Ruby and Rust then it may be good to work backwards to find what was the breaking change (this is the most tedious route IMO).  The other is that since all the errors I've gotten seem to be related to only one method then I can try rewriting Ruby's internal id lookup in Rust and have Windows use that... in other words simply sidestep the issue.  I'm not sure when I'd have time to take on this task but this is the strategy I've thought of.

**Comment by pravic** [comment]
Yes, I tested on the latest Ruby version. Is there any other known-good version instead?

**Comment by danielpclark** [comment]
Here's the `.appveyor.yml` file from RuRu https://github.com/d-unseductable/ruru/blob/master/.appveyor.yml .  It worked with Ruby 2.4 with rust's gnu toolchain for Rust 1.16.  Rutie has since added features requiring later versions of Rust so rolling back Rust versions would be difficult.

**Comment by pravic** [comment]
No, it's not about Rust version. It's about compatibility with the MSVC runtime (CRT).
